tchannel-core/src/main/java/com/uber/tchannel/tracing/PrefixedHeadersCarrier.java

```diff
@@ -0,0 +1,124 @@
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+
+package com.uber.tchannel.tracing;
+
+import com.google.common.base.Function;
+import com.google.common.base.Predicate;
+import com.google.common.collect.Iterators;
+import com.google.common.collect.Maps;
+import io.opentracing.propagation.TextMap;
+
+import java.util.Iterator;
+import java.util.Map;
+
+class PrefixedHeadersCarrier implements TextMap {
+    /**
+     * Prefix used with all keys generated by the tracer when marshalling SpanContext
+     * to a map of headers.
+     * <p>
+     * This string value is shared across all implementations of TChannel.
+     */
+    private static final String TRACING_KEY_PREFIX = "$tracing$";
+
+    private static final Function<String, String> PREFIXED_KEYS = prefixedKeys(TRACING_KEY_PREFIX);
+
+    private static final Function<String, String> UNPREFIXED_KEYS = unprefixedKeys(TRACING_KEY_PREFIX);
+
+    private final Map<String, String> headers;
+    private final String prefix;
+    private final Function<String, String> encoder;
+    private final Function<String, String> decoder;
+
+    PrefixedHeadersCarrier(Map<String, String> headers) {
+        this(headers, TRACING_KEY_PREFIX, PREFIXED_KEYS, UNPREFIXED_KEYS);
+    }
+
+    PrefixedHeadersCarrier(Map<String, String> headers, String prefix) {
+        this(headers, prefix, prefixedKeys(prefix), unprefixedKeys(prefix));
+    }
+
+    private PrefixedHeadersCarrier(
+            Map<String, String> headers,
+            String prefix,
+            Function<String, String> encoder,
+            Function<String, String> decoder
+    ) {
+        this.headers = headers;
+        this.prefix = prefix;
+        this.encoder = encoder;
+        this.decoder = decoder;
+    }
+
+    @Override
+    public Iterator<Map.Entry<String, String>> iterator() {
+        return Iterators.transform(Iterators.filter(headers.entrySet().iterator(),
+                new Predicate<Map.Entry<String, String>>() {
+                    @Override
+                    public boolean apply(Map.Entry<String, String> entry) {
+                        return entry.getKey().startsWith(prefix);
+                    }
+                }),
+                new Function<Map.Entry<String, String>, Map.Entry<String, String>>() {
+                    @Override
+                    public Map.Entry<String, String> apply(Map.Entry<String, String> entry) {
+                        return Maps.immutableEntry(
+                                decoder.apply(entry.getKey()),
+                                entry.getValue());
+                    }
+                });
```

Comment:
I like Guava's functional primitives as much as the next person, but I think
this code would be more readable without the `Predicate` and `Function` noise.
Even the authors of Guava have something similar to say on this: https://github.com/google/guava/wiki/FunctionalExplained#caveats

Consider using one of the following,

- Create an `AbstractIterator` (provided by Guava) inline:

    ```python
    final Iterator<Map.Entry<String, String>> headers = ...;
    
    return new AbstractIterator<Map.Entry<String, String>>() {
        protected Map.Entry<String, String> computeNext() {
            while (in.hasNext()) {
                Map.Entry<String, String> entry = in.next();
                if (!entry.getKey().startsWith(prefix)) {
                    continue;
                }
                
                return Maps.immutableEntry(
                    decoder.apply(entry.getKey()),
                    entry.getValue(),
                );
            }
            return this.endOfData();
        }
    }
    ```

- Create a private inner class.
    
    ```java
    @Override
    public Iterator<Map.Entry<String, String>> iterator() {
        return new PrefixIterator(headers.entrySet().iterator());
    }
    
    private class PrefixIterator extends AbstractIterator<...> {
        // ...
    }
    ```

- Move the construction of the `Function` and `Predicate` into helper
  functions.

    ```java
    @Override
    public Iterator<Map.Entry<String, String>> iterator() {
        return Iterators.transform(
            Iterators.filter(headers.entrySet.iterator(), prefixMatcher),
            keyDecoder,
        );
    }
    ```